pulp/apis/coin_api.py

```diff
@@ -176,31 +178,70 @@ def solve_CBC(self, lp, use_mps=True):
             cmds += "initialSolve "
         cmds += "printingOptions all "
         cmds += "solution " + tmpSol + " "
-        if self.msg:
-            pipe = None
-        else:
-            pipe = open(os.devnull, "w")
+
         logPath = self.optionsDict.get("logPath")
-        if logPath:
-            if self.msg:
-                warnings.warn(
-                    "`logPath` argument replaces `msg=1`. The output will be redirected to the log file."
-                )
-            pipe = open(self.optionsDict["logPath"], "w")
+
         log.debug(self.path + cmds)
         args = []
         args.append(self.path)
         args.extend(cmds[1:].split())
-        cbc = subprocess.Popen(args, stdout=pipe, stderr=pipe, stdin=devnull)
-        if cbc.wait() != 0:
-            if pipe:
-                pipe.close()
+
+        class SubProcessRunner:
```

Comment:
I'm not the maintainer of the project so this is just an opinion, but to make testing easier, this class could be split out at a higher level and tested independently, rather than having the class declared inline.